Scope Databricks inferred pip requirements to modules

dbczumar opened the pull request:
Signed-off-by: dbczumar <corey.zumar@databricks.com>

## What changes are proposed in this pull request?

Scope Databricks inferred pip requirements to modules

## How is this patch tested?

- Added unit tests
- Tested manually on Databricks:
```
import mlflow.pyfunc


import logging
logging.getLogger("mlflow").setLevel(logging.DEBUG)

class MyModel(mlflow.pyfunc.PythonModel):
  def load_context(self, context):
    import sklearn
    import databricks.model_monitoring
    
  def predict(self, ctx, inp):
    return 7
  
with mlflow.start_run():
  model = mlflow.pyfunc.log_model(
    python_model=MyModel(),
    artifact_path="model"
  )

import mlflow.models

mlflow.models.infer_pip_requirements(model.model_uri, "python_function")
```

```
Out[2]: ['configparser==5.0.1',
 'databricks-model-monitoring==0.0.2',
 'psutil==5.8.0',
 'pyspark==3.2.1',
 'scikit-learn==0.24.1']
```

```
import mlflow.pyfunc


import logging
logging.getLogger("mlflow").setLevel(logging.DEBUG)

class MyModel(mlflow.pyfunc.PythonModel):
  def load_context(self, context):
    import sklearn
    import databricks.model_monitoring
    import databricks.automl
    
  def predict(self, ctx, inp):
    return 7
  
with mlflow.start_run():
  model = mlflow.pyfunc.log_model(
    python_model=MyModel(),
    artifact_path="model"
  )

import mlflow.models

mlflow.models.infer_pip_requirements(model.model_uri, "python_function")
```

```
Out[4]: ['configparser==5.0.1',
 'databricks-automl-runtime==0.2.7',
 'databricks-model-monitoring==0.0.2',
 'fasttext==0.9.2',
 'future==0.18.2',
 'holidays==0.13',
 'networkx==2.5',
 'notebook==6.3.0',
 'psutil==5.8.0',
 'pyspark==3.2.1',
 'scikit-learn==0.24.1',
 'simplejson==3.17.2',
 'tqdm==4.59.0',
 'wrapt==1.12.1']
```

```
import mlflow.pyfunc


import logging
logging.getLogger("mlflow").setLevel(logging.DEBUG)

class MyModel(mlflow.pyfunc.PythonModel):
  def load_context(self, context):
    import sklearn
    import databricks.model_monitoring
    import databricks.automl
    import databricks.koalas
    
  def predict(self, ctx, inp):
    return 7
  
with mlflow.start_run():
  model = mlflow.pyfunc.log_model(
    python_model=MyModel(),
    artifact_path="model"
  )

import mlflow.models

mlflow.models.infer_pip_requirements(model.model_uri, "python_function")
```

# PySpark already includes koalas, so the set of dependencies doesn't contain a koalas entry:
```
Out[6]: ['configparser==5.0.1',
 'databricks-automl-runtime==0.2.7',
 'databricks-model-monitoring==0.0.2',
 'fasttext==0.9.2',
 'future==0.18.2',
 'holidays==0.13',
 'networkx==2.5',
 'notebook==6.3.0',
 'psutil==5.8.0',
 'pyspark==3.2.1',
 'scikit-learn==0.24.1',
 'simplejson==3.17.2',
 'tqdm==4.59.0',
 'wrapt==1.12.1']
```

```
WARNING:root:Found pyspark version "3.2.1" installed. The pyspark version 3.2 and above has a built-in "pandas APIs on Spark" module ported from Koalas. Try `import pyspark.pandas as ps` instead. 
```

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

## Release Notes

When inferring model requirements, only include Databricks libraries if their corresponding modules are imported by the model.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

dbczumar committed:
Fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Format

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy commented:
I'm wondering why `koalas` is included in the second test example 🤔

```python
import mlflow.pyfunc


import logging
logging.getLogger("mlflow").setLevel(logging.DEBUG)

class MyModel(mlflow.pyfunc.PythonModel):
  def load_context(self, context):
    import sklearn
    import databricks.model_monitoring
    import databricks.automl
    
  def predict(self, ctx, inp):
    return 7
  
with mlflow.start_run():
  model = mlflow.pyfunc.log_model(
    python_model=MyModel(),
    artifact_path="model"
  )

import mlflow.models

mlflow.models.infer_pip_requirements(model.model_uri, "python_function")

```

```
Out[8]: ['configparser==5.0.1',
 'databricks-automl-runtime==0.2.7',
 'databricks-model-monitoring==0.0.2',
 'fasttext==0.9.2',
 'future==0.18.2',
 'holidays==0.13',
 'koalas==1.8.2',
 'networkx==2.5',
 'notebook==6.3.0',
 'psutil==5.8.0',
 'pyspark==3.2.1',
 'scikit-learn==0.24.1',
 'simplejson==3.17.2',
 'tqdm==4.59.0',
 'wrapt==1.12.1']
```

harupy commented on mlflow/utils/requirements_utils.py:
```diff
@@ -281,6 +281,10 @@ def _capture_imported_modules(model_uri, flavor):
 
+DATABRICKS_MODULES_TO_PACKAGES = {
+    "databricks.automl": ["databricks-automl-runtime"],
+    "databricks.model_monitoring": ["databricks-model-monitoring"],
```
Can Model Serving serve a model that depends on `databricks-model-monitoring` (which is not available on PyPI)?

harupy replied:
Usually, a model doesn't depend on `databricks.model_monitoring`, right?

dbczumar replied:
Correct, a model doesn't usually depend on this, but Serving folks are adding support for it just in case.

dbczumar committed:
Fix attempt

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented:
@mlflow-automation autoformat

dbczumar committed:
Print

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Comment

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Debug

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Explicit mapping of automl_runtime

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Test fixes and format

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on tests/utils/test_requirements_utils.py:
```diff
@@ -315,3 +316,60 @@ def test_infer_requirements_does_not_print_warning_for_recognized_packages():
         mock_warning.assert_not_called()
+
+
+def test_capture_imported_modules_scopes_databricks_imports(request, tmpdir):
```
Thanks! Applied!

dbczumar commented:
> I'm wondering why `koalas` is included in the second test example 🤔
> 
> ```python
> import mlflow.pyfunc
> 
> 
> import logging
> logging.getLogger("mlflow").setLevel(logging.DEBUG)
> 
> class MyModel(mlflow.pyfunc.PythonModel):
>   def load_context(self, context):
>     import sklearn
>     import databricks.model_monitoring
>     import databricks.automl
>     
>   def predict(self, ctx, inp):
>     return 7
>   
> with mlflow.start_run():
>   model = mlflow.pyfunc.log_model(
>     python_model=MyModel(),
>     artifact_path="model"
>   )
> 
> import mlflow.models
> 
> mlflow.models.infer_pip_requirements(model.model_uri, "python_function")
> ```
> 
> ```
> Out[8]: ['configparser==5.0.1',
>  'databricks-automl-runtime==0.2.7',
>  'databricks-model-monitoring==0.0.2',
>  'fasttext==0.9.2',
>  'future==0.18.2',
>  'holidays==0.13',
>  'koalas==1.8.2',
>  'networkx==2.5',
>  'notebook==6.3.0',
>  'psutil==5.8.0',
>  'pyspark==3.2.1',
>  'scikit-learn==0.24.1',
>  'simplejson==3.17.2',
>  'tqdm==4.59.0',
>  'wrapt==1.12.1']
> ```

Great catch! Turns out `databricks.automl_runtime` is also being imported, which isn't in the hardcoded module list. I've added some extra logic so that `databricks.automl*` maps to the automl package unless there's a separate package that matches the second-level module name precisely. I've updated the PR description to reflect updated testing results.

dbczumar committed:
Fix tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy reviewed (approved):
LGTM!
